### PR TITLE
http3: support Http3Options for downstream 

### PR DESCRIPTION
**Pull request by danzh2010**
Commit Message: allow HCM to config Http3Options and use it with other HCM configs,  i.e. max_request_headers_kb and headers_with_underscores_action, to setup QuicHttpServerConnectionImpl. And support these configurations in QUIC. Currently the only Http3Options configuration is override_stream_error_on_invalid_http_message.

Additional Description: added  Http3 codec stats. Pass it along with Http3Options and other HCM configs.

Risk Level: low 
Testing: enable related tests in quic_protocol_integration_test
Docs Changes: updated docs/root/configuration/http/http_conn_man/response_code_details.rst

Part of #12930 #2557

### CODE REVIEW

File: source/common/quic/envoy_quic_server_stream.cc
Comment (alyssawilk) [inline source/common/quic/envoy_quic_server_stream.cc:357]: can we factor the Http2 code into HeaderUtility or not enough overlap?  or maybe at least stick it in the utility directory, unit terst, and we can use when we switch codec libraries.

File: source/common/quic/envoy_quic_server_stream.h
Comment (alyssawilk) [inline source/common/quic/envoy_quic_server_stream.h:67]: ?

File: source/common/quic/envoy_quic_stream.h
Comment (alyssawilk) [inline source/common/quic/envoy_quic_stream.h:115]: Ditto, consider moving to utility

**Review by alyssawilk — COMMENTED** [review]
Nice change!

Once you sort out docs CI, let's check coverage on this one and make sure we've got corner cases tested.

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto
Comment (danzh2010) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:330]: @htuch The doc CI fails because this field is annotated with udpa.annotations.security, but doesn't have description in docs/protodoc_manifest.yaml. Do I need to document override_stream_error_on_invalid_http_message, the only field of Http3ProtocolOptions, there? http2_protocol_options doesn't have documentation about its similar field in protodoc_manifest.yaml.

**Comment by alyssawilk** [comment]
Try hiding the field and see if that fixes docs build - it's probably trying to generate docs for the (not hidden) field refering to the (hidden) proto and having problems.
Also please main merge to pick up the Matt moving-all-the-things PR

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto
Comment (htuch) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto]: If there is anything security related, e.g. buffering, update the example manifest YAML as per offline discussion. If not, feel free to drop the security annotation.

File: api/envoy/config/core/v3/protocol.proto
Comment (alyssawilk) [reply api/envoy/config/core/v3/protocol.proto:429]: http2 has an override to allow different behavior for HTTP/1 and HTTP/2.  We have regression tests for being able to do protocol-specific overrides which we're trying to turn up for HTTP/3.

File: source/common/quic/envoy_quic_server_stream.cc
Comment (danzh2010) [reply source/common/quic/envoy_quic_server_stream.cc]: I moved this method to HeaderUtility, but using it in h2 codec doesn't save many lines, so I didn't use it there.

File: source/common/quic/envoy_quic_stream.h
Comment (danzh2010) [reply source/common/quic/envoy_quic_stream.h]: done

File: source/common/quic/envoy_quic_server_stream.h
Comment (danzh2010) [reply source/common/quic/envoy_quic_server_stream.h:67]: It was from the stream buffer PR. Merged now.

File: api/envoy/config/core/v3/protocol.proto
Comment (htuch) [reply api/envoy/config/core/v3/protocol.proto:429]: Fair

**Review by htuch — COMMENTED** [review]
/lgtm api

File: source/common/http/header_utility.h
Comment (alyssawilk) [inline source/common/http/header_utility.h:221]: optional nit: generally we don't do | around names

File: source/common/quic/envoy_quic_stream.h
Comment (alyssawilk) [inline source/common/quic/envoy_quic_stream.h:113]: Given we're cloning this, what do you think of just always setting it in initializeAndValidate?

File: source/common/quic/quic_filter_manager_connection_impl.h
Comment (alyssawilk) [inline source/common/quic/quic_filter_manager_connection_impl.h:167]: typo?

**Review by alyssawilk — COMMENTED** [review]
Cool, just a few more nits and I think this one's good to go

**Review by alyssawilk — APPROVED** [review]

File: source/common/quic/envoy_quic_server_stream.cc
Comment (mattklein123) [inline source/common/quic/envoy_quic_server_stream.cc:42]: nit: it's kind of strange that this one is uses dot separation and the others are using underscores? Can we be consistent? Is this copied from elsewhere?

File: include/envoy/upstream/upstream.h
Comment (mattklein123) [inline include/envoy/upstream/upstream.h:999]:
```suggestion
   * @return the Http3 Codec Stats.
```

File: source/common/quic/envoy_quic_server_session.cc
Comment (mattklein123) [inline source/common/quic/envoy_quic_server_session.cc:49]: Can this actually happen? If it can, can this log spam?

File: source/common/http/header_utility.cc
Comment (mattklein123) [inline source/common/http/header_utility.cc:336]: It kills me that this code is now copied into a third place when we already effectively have it in both the H1 and H2 codec. This is really scary and security sensitive stuff and I feel we shouldn't be making the problem worse (it's in the opposite direction of https://github.com/envoyproxy/envoy/issues/10646). Is there any work we can do to share some of this code now? Maybe we can do a different PR to add this helper and call it somehow from both other codecs?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks really awesome to see this all coming together. A few comments from a quick pass.

/wait

File: source/common/http/header_utility.cc
Comment (danzh2010) [reply source/common/http/header_utility.cc:336]: I don't want to touch any code in production in this PR because that needs feature protection. Once this PR lands, the other two codec can switch to this utility function.

File: source/common/quic/envoy_quic_server_stream.cc
Comment (danzh2010) [reply source/common/quic/envoy_quic_server_stream.cc:42]: I saw `http2.invalid.header.field` and thought any error from external library should use `.`, no?

File: source/common/quic/envoy_quic_server_session.cc
Comment (mattklein123) [reply source/common/quic/envoy_quic_server_session.cc:49]: OK then please use either ASSERT or ENVOY_BUG depending on what you want to do in terms of providing testing. I don't think kind of error handling makes sense.

File: source/common/quic/envoy_quic_server_stream.cc
Comment (mattklein123) [reply source/common/quic/envoy_quic_server_stream.cc:42]: I think this is probably just a typo. IMO I would make it consistent with the others.

File: source/common/http/header_utility.cc
Comment (mattklein123) [reply source/common/http/header_utility.cc:336]: OK fair enough. Can you at least add comments and put TODOs in the other codec places which reference this code and the tracking issue for merging?

**Comment by danzh2010** [comment]
I refactored the stream error handling a bit. PTAL

File: source/common/quic/envoy_quic_server_session.cc
Comment (mattklein123) [inline source/common/quic/envoy_quic_server_session.cc:47]: Do you have a test for this ENVOY_BUG? If not just make it an assert.

**Review by mattklein123 — DISMISSED** [review]
Thanks LGTM, will defer to @alyssawilk to review the most recent set of changes and merge. Thank you!

File: source/common/quic/envoy_quic_server_session.cc
Comment (danzh2010) [reply source/common/quic/envoy_quic_server_session.cc:47]: As long as our implementation is correct, it shouldn't be reached. So I don't know how to trigger it in a meaningful test right now. But I'd like to keep it in production, in case we miss some QUICHE behavior change that could occassionally lead to this in the future.

File: source/common/http/header_utility.cc
Comment (alyssawilk) [reply source/common/http/header_utility.cc:336]: cc @yanavlasov might be a good starter project for one of the envoy-sec folks

File: source/common/quic/envoy_quic_server_stream.cc
Comment (alyssawilk) [inline source/common/quic/envoy_quic_server_stream.cc:167]: what do you think of making this a default argument?  If you'd prefer to be explicit I'd mildly prefer a helper function for brevity.

**Review by alyssawilk — COMMENTED** [review]
Nice!  just one more nit and good to go.

File: source/common/quic/envoy_quic_server_session.cc
Comment (mattklein123) [reply source/common/quic/envoy_quic_server_session.cc:47]: But how will you know the error checking will work if we reach it in production? :) In general we require testing for ENVOY_BUG type logic, so I would still prefer that you switch this to assert if you can't reasonably test it.

Comment (danzh2010) [reply source/common/quic/envoy_quic_server_session.cc:47]: Okay, I added a unit test in envoy_quic_server_session_test to hit this ENVOY_BUG just to check the ENVOY_BUG works as intended in debug and release build.

File: source/common/quic/envoy_quic_server_stream.cc
Comment (danzh2010) [reply source/common/quic/envoy_quic_server_stream.cc:167]: I'd prefer having explicit argument. I changed the function to take an optional bool which if absent means checking override_stream_error_on_invalid_http_message from h3 options.

File: source/common/quic/envoy_quic_server_stream.h
Comment (alyssawilk) [inline source/common/quic/envoy_quic_server_stream.h:94]: update comment, please?

Comment (danzh2010) [reply source/common/quic/envoy_quic_server_stream.h:94]: done

**Comment by alyssawilk** [comment]
API hasn't changed since the LGTM so I'm going to go ahead and merge.

**Pull request merged**
